source/common/http/http1/codec_impl.cc

```diff
@@ -101,20 +104,12 @@ StreamEncoderImpl::StreamEncoderImpl(ConnectionImpl& connection,
   }
 }
 
-void StreamEncoderImpl::encodeHeader(const char* key, uint32_t key_size, const char* value,
-                                     uint32_t value_size) {
-
-  ASSERT(key_size > 0);
-  const uint64_t old_buffer_length = connection_.buffer().length();
-  connection_.copyToBuffer(key, key_size);
-  connection_.addCharToBuffer(':');
-  connection_.addCharToBuffer(' ');
-  connection_.copyToBuffer(value, value_size);
-  connection_.addToBuffer(CRLF);
-  bytes_meter_->addHeaderBytesSent(connection_.buffer().length() - old_buffer_length);
-}
 void StreamEncoderImpl::encodeHeader(absl::string_view key, absl::string_view value) {
```

Comment:
I wonder if we could stream line in 
https://github.com/envoyproxy/envoy/pull/19115/files#diff-8d3b9d0a839b05612561930ee8c58bf1b0ff0adc55cf27d79a7e3bbfbd81965eR154

such that we create storage for the entire header map serialized and write that instead to the connection buffer?

Reply:
@KBaichoo I have implemeted a helper class which will reserve a long continuous memory (4K+) from buffer. And when the local reservation is overflow then the helper class will commit this reservation to the buffer. It is my previous version and you can check the code from https://github.com/envoyproxy/envoy/pull/19115/commits/121f9123cbbdcaaa875c0fc343db15c3dcf64dd1.

Do you mean something like this?

Reply:
Yes, similar to that. Basically 121f912 with the initial reservation being the size of the header map, and extra space for colon, space, CRLF.  We'd only commit once for the headermap (checking the high watermark once for header map vs number of header times as done in fragments).

Reply:
@KBaichoo Thanks for you suggestion. If we directly reserve the size of the header map, we can get better performance. 

The latest implementation of the API is simpler, and the previous implementation will have better performance, and it is an independent class, which can be more flexible.
I don’t know what the two cc @jmarantz @antoniovicente 's opinions are. I personally think that both have their value. It is hard to decide.

Reply:
Are there other places we'd use `addFragments`?

The reservation for the entire headermap would couple with the specific H1 codec, but we don't write out headers one by one in the other codecs (e.g. in H2 we would write out a header frame). So I think H1 is really where we have this problem.

Reply:
@KBaichoo Currently only h1 codec uses this method. For me, reverting to the previous implementation is very simple. But the community needs to agree on this. 
So, can @jmarantz @antoniovicente give this some more suggestions?